std/net/curl.d

```diff
@@ -186,16 +186,95 @@ version(unittest)
     import std.file : tempDir;
     import std.path : buildPath;
 
-    // Source code for the test service is available at
-    // https://github.com/jcd/d-lang-testservice
-    enum testService = "d-lang.appspot.com";
-
-    enum testUrl1 = "http://"~testService~"/testUrl1";
-    enum testUrl2 = "http://"~testService~"/testUrl2";
-    // No anonymous DigitalMars FTP access as of 2015
-    //enum testUrl3 = "ftp://ftp.digitalmars.com/sieve.ds";
-    enum testUrl4 = testService~"/testUrl1";
-    enum testUrl5 = "http://"~testService~"/testUrl3";
+    import std.socket : Address, INADDR_LOOPBACK, Socket, TcpSocket;
+
+    private struct TestServer
+    {
+        string addr() { return _addr; }
+
+        void handle(void function(Socket s) dg)
+        {
+            tid.send(dg);
+        }
+
+    private:
+        string _addr;
+        Tid tid;
+
+        static void loop(shared TcpSocket listener)
+        {
+            try while (true)
+            {
+                void function(Socket) handler = void;
+                try
+                    handler = receiveOnly!(typeof(handler));
+                catch (OwnerTerminated)
+                    return;
+                handler((cast()listener).accept);
+            }
+            catch (Throwable e)
+            {
+                import core.stdc.stdlib : exit, EXIT_FAILURE;
+                stderr.writeln(e);
+                exit(EXIT_FAILURE); // Bugzilla 7018
+            }
+        }
+    }
+
+    private TestServer startServer()
+    {
+        auto sock = new TcpSocket;
+        sock.bind(new InternetAddress(INADDR_LOOPBACK, InternetAddress.PORT_ANY));
+        sock.listen(1);
+        auto addr = sock.localAddress.toString();
+        auto tid = spawn(&TestServer.loop, cast(shared)sock);
+        return TestServer(addr, tid);
+    }
+
+    private ref TestServer testServer()
+    {
+        __gshared TestServer server;
+        return initOnce!server(startServer());
+    }
+
+    private immutable(T)[] recvAll(T=char)(Socket s)
+    {
+        ubyte[1024] buf=void;
+        ubyte[] res;
+        while (true)
+        {
+            auto nbytes = s.receive(buf[]);
+            assert(nbytes >= 0);
+            res ~= buf[0 .. nbytes];
+            if (nbytes < buf.length)
```

Comment:
Technically short read doesn't mean no more data - it just means we are reading faster then data comes in (split in whatever bits and pieces).

recvAll inevitably has to including some framing logic unless we want to get bugs down the road. Try to detect ` Content-Length: [0-9]+ \r\n` ?